File: docs/src/neetcode150/go/valid_sudoku.go
package main

func isValidSudoku(board [][]byte) bool {
	var rows, cols, boxes [9][9]bool

	for r := range board {
		for c := range board[0] {
			cell := board[r][c]
			if cell != '.' {
				b := (r/3)*3 + (c / 3)
				idx := int(cell - '1')
				if rows[r][idx] || cols[c][idx] || boxes[b][idx] {
					return false
				}
				rows[r][idx] = true
				cols[c][idx] = true
				boxes[b][idx] = true
			}
		}
	}
	return true
}

File: docs/DSA/src/py/valid_sudoku.py
from collections import defaultdict
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = defaultdict(set)
        cols = defaultdict(set)
        box = defaultdict(set)
        for r in range(len(board)):
            for c in range(len(board[0])):
                cell = board[r][c]
                b = (r // 3) * 3 + c // 3
                if cell == ".":
                    continue
                elif cell in rows[r] or cell in cols[c] or cell in box[b]:
                    return False

                rows[r].add(cell)
                cols[c].add(cell)
                box[b].add(cell)
        return True

